src/python/pants/console/stty_utils.py

```diff
@@ -5,33 +5,43 @@
 from __future__ import (absolute_import, division, generators, nested_scopes, print_function,
                         unicode_literals, with_statement)
 
+import logging
+import sys
+import termios
+import tty
 from contextlib import contextmanager
 
-from pants.util.process_handler import subprocess
 
-
-@contextmanager
-def preserve_stty_settings():
-  """Run potentially stty-modifying operations, e.g., REPL execution, in this contextmanager."""
-  stty_settings = STTYSettings()
-  stty_settings.save_stty_options()
-  yield
-  stty_settings.restore_ssty_options()
+logger = logging.getLogger(__name__)
 
 
 class STTYSettings(object):
   """Saves/restores stty settings, e.g., during REPL execution."""
 
-  def __init__(self):
-    self._stty_options = None
-
-  def save_stty_options(self):
-    self._stty_options = self._run_cmd('stty -g 2>/dev/null')
+  @classmethod
+  @contextmanager
+  def preserved(cls):
+    """Run potentially stty-modifying operations, e.g., REPL execution, in this contextmanager."""
+    inst = cls()
+    inst.save_tty_flags()
+    try:
+      yield
+    finally:
+      inst.restore_tty_flags()
 
-  def restore_ssty_options(self):
-    self._run_cmd('stty ' + self._stty_options)
-
-  def _run_cmd(self, cmd):
-    po = subprocess.Popen(cmd, shell=True, stdout=subprocess.PIPE)
-    stdout, _ = po.communicate()
-    return stdout
+  def __init__(self):
+    self._tty_flags = None
+
+  def save_tty_flags(self):
+    # N.B. `stty(1)` operates against stdin.
+    try:
+      self._tty_flags = tty.tcgetattr(sys.stdin.fileno())
```

Comment:
Is accessing `termios` module functions and constants via the `tty` module idiomatic? I don't know this space; so it is a bit confusing.

Reply:
ah, good call - fixed.